### ISSUE TITLE
Not seeing or hearing Near sweater or Near wall landmarks when dragging balloon

### ISSUE THREAD

**Issue by terracoda** [phetsims/balloons-and-static-electricity#313]
In version 1.4.0-dev.4, I do not see the Near sweater and Near wall landmarks. A crossed landmark should always be spoken when moving the balloon.

I see the landmarks if I use the jump-to hot keys, but during regular moving towards sweater or towards wall.

A user should not have to stop right on these 2 landmarks to hear them, but rather just cross them while going in the right direction. Does that make sense?

@jessegreenberg, perhaps this is on the todo list still?

**Comment by terracoda** [comment]
These landmarks are described under the heading [Dragging Alert 1: Landmark and Special Landmark Identification](https://docs.google.com/document/d/1WNVq5_-Hc9LSdeZVSn2edlwYwpHtGNT1ThFD9NXpZ5Q/edit#heading=h.h2vv8qkksmr4) in the design document.

**Comment by terracoda** [comment]
Making sure these 2 landmarks are announced would possibly fill in some of the blank steps that were experienced recently by screen reader users.

**Comment by terracoda** [comment]
@jessegreenberg, the screen reader users got way more blank/silent steps than I am getting, so I don't know how they are pressing their keyboard.

I can get 3 blank steps and very rarely 4 when using the arrow keys. 
When using WASD keys, I generally only get single blank steps. 

However, I did notice that while using Voice Over, the left and right arrow keys can get sluggish. The up and down arrow keys seem more resilient. A sim reset usually fixes the sluggish arrow keys. Also, dragging alerts can just stop after a lot of use. And when that happens the sim needs to be reloaded to get dragging alerts going again.

**Comment by terracoda** [comment]
Considering previous https://github.com/phetsims/balloons-and-static-electricity/issues/313#issuecomment-334275824 ensuring the "Near sweater" and "Near wall" landmarks are firing will greatly reduce the number of blank steps.

I am trying not to think of using key presses... With that said, I am wondering if the first step just mentions direction, then the second step indicates position, we might get fewer blank steps.

It seems to me that drag velocity is much less relevant when using a keyboard, and the distance through the grid cell when using a keyboard is what matters. 

What's the current algorithm?
Are we using what I described in my pseudo logic in the design document?

**Comment by jessegreenberg** [comment]
Thanks @terracoda, verified, Im looking into it now.

> I am wondering if the first step just mentions direction, then the second step indicates position, we might get fewer blank steps.

In the sim, there is no such thing as one 'step'.  When the balloon moves, it moves *continuously* (maybe a pixel or two with good performance). How far the balloon moves is dependent on how long a key is held down in milliseconds.  I think that thinking in terms of steps would require a change to the entire interaction pattern for the balloon.

> Are we using what I described in my pseudo logic in the design document?

That is the goal, but there are issues (including this one).

**Comment by jessegreenberg** [comment]
Alright, landmarks should be in there now, but it would be great for @terracoda  to verify in the next version so makring as "pending testing" and will reassign once next version is ready.

**Comment by terracoda** [comment]
@jessegreenberg, thanks! And I agree about not thinking about it the interaction in steps. I am just having trouble understanding how keyboard presses relate to the drag velocity property. 

Looking forward to hearing the new version :-)

**Comment by terracoda** [comment]
@jessegreenberg, maybe this should be a new issue? Not sure. It is related to reducing the number of blank steps.

I did some more testing today of Dev Version 1.4.0-dev.4 (without near sweater and near wall) and having these landmarks will definitely reduce number of blank steps.

Other things I noticed (Using A key only) are:
- Even when trying I never get more than 3 blank steps in a row between sweater and wall.
- I always get a blank step on key press 2 when starting from center of Play Area.
- When using the shift key I get significantly more blank steps (up to 8) between "On left side of Play Area" and "Closer to Sweater". Going slow is definitely a problem.

In the design doc, I had a few special landmarks (On sweater, Off sweater, Very close to sweater, Very close to wall). On and Off sweater are fine, but I never hear the "very close to" special landmark. We don't necessarily need to add that special landmark, but may need it for when users are going slow (i.e., shift-key velocity).

Currently, the algorithm is:
- Repeat actual location as long as drag velocity is normal (greater than shift-key velocity), horizontal movement is relatively steady, AND distance through grid cell is less than 66%.

I think we need to decrease distance through grid cell, possibly to as low as 30%, but we may not need it that low once we have all the landmarks being identified. A smaller threshold should increase the number of utterances of "actual locations" and eliminate blank steps for "normal velocity", and hopefully not be too repetitive.

For a slow velocity, it seems more problematic (i.e., currently many blank steps). We might need to decrease distance through grid cell to something like 10%, which would increase the number of Alert 4's (Closer to destination), and we may need to add in the special landmarks, "Very close to sweater" and "Very close to wall". 

Or we just might have provide a hint to move faster.

**Comment by terracoda** [comment]
@jessegreenberg, I just checked the built version for the first time. I didn't realize that was how to get the latest version. Sorry about that.

I hear the On Center of Play Area, Near Sweater, Near Wall, At Wall, and Off Sweater all at the right times, but now I do not hear the actual locations, "On left side of Play Area" and "On right side of Play Area."

We need to hear Landmarks, Actual Locations, and Special Landmarks when dragging in order to figure out what percentage through the grid cell is ideal for reducing the number of blank steps.

**Comment by terracoda** [comment]
@jessegreenberg, as of last Friday, the landmarks were working fine. You can likely close this issue.

On a related note, I am just wondering if you changed the algorithm for distance through the grid cell at all, or if you just managed to get all the locations and landmarks working properly. 

When I tested last Friday, I only ever got a single blank step here and there. Over all dragging alerts in the open play area sounded good to me. 

Once you get the sweater dragging alerts in. I can retest with Screen Reader users, and see if we need to adjust the algorithm for distance through the grid cell to try to eliminate the blank steps.

**Comment by jessegreenberg** [comment]
@terracoda improvements to behavior mostly came from major changes to the timing of alerts, but the algorithm should also match what you have in BASE: Grid Descriptions Simplified and Revised design document now.
